### PR TITLE
fix(cdaas-2561): start command posix issues

### PR DESCRIPTION
**Pull request by fieldju**
## Motivation and Context
Please note that this is a surgical fix, this fix layouts the ground work for fixing our use of exit codes and stderr and stdout in general.

This fix is only aims at the start deployment command.
We are making it not swallow errors but rather return them from the `RunE` command and then in the main method we can log the errors to `stderr` and exit with the appropriate code.

This is surgical fix is: [CDAAS-2561]

Applying the patterns in this PR to the rest of the commands would solve:
- [CDAAS-1786]
- [CDAAS-2585]

This PR enables bash scripts around our CLI to work as expected such as

```bash
#!/usr/bin/env bash

set -e
set -o pipefail

DEPLOYMENT_ID=$(armory deploy start -f scratch/deployment.yaml -o json | jq -r ".deploymentId")
DEPLOYMENT_START_EXIT_CODE=$?

// if $DEPLOYMENT_START_EXIT_CODE == 0 sucess, $DEPLOYMENT_ID will be a valid deployment id
// if $DEPLOYMENT_START_EXIT_CODE == 3 confict, deployment already in progress
// if $DEPLOYMENT_START_EXIT_CODE == 1 some other error
```

## How has this been tested? How can a reviewer test these changes?

I manually tested it with the included run.sh and debug.sh scripts

image
image
image

I also updated the unit tests and added a new test to ensure a 409 conflict has an exit code 3 when started a deployments

# Have your performed the following tests?

Please confirm you have done the following tests to ensure your pull request is ready to be merged..

- [X] If this change modifies the deployment flow have triggered a deployment using this branch (i.e. `build/dist/darwin_amd64/armory deploy start -f <path to file>`)?
- [X] Have you verified the specific change made in this PR?

[CDAAS-2561]: https://armory.atlassian.net/browse/CDAAS-2561?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CDAAS-1786]: https://armory.atlassian.net/browse/CDAAS-1786?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CDAAS-2585]: https://armory.atlassian.net/browse/CDAAS-2585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/cluster/sandbox_file_writer.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func (d *SandboxClusterFileStore) getSandboxFileLocation() (string, error) {
 	var dirname string
 	var err error
 	if isATest {
-		return os.TempDir() + "dotarmory_sandbox", nil
+		return os.TempDir() + "/.armory_sandbox", nil
```
Comment (fieldju) [inline cmd/cluster/sandbox_file_writer.go:77]: Test wasn't passing on my machine without the `/`

Comment (Greg-At-Armory) [reply cmd/cluster/sandbox_file_writer.go:77]: i did get your branch on my local machine and executed tests - no problems with execution, so i guess this is safe for everyone

File: internal/clierr/api_error.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+// Error the string representing the error.
+func (a *APIError) Error() string {
+	return fmt.Sprintf("an API error occurred, msg: %v, status code: %v", a.message, a.httpStatusCode)
+}
+
+// DetailedError returns a human readable multiline colored error message.
+func (a *APIError) DetailedError() string {
+	msg := color.New(color.FgRed, color.Bold).Sprintln(a.message)
```
Comment (Greg-At-Armory) [inline internal/clierr/api_error.go:45]: i'm curious - would it work with various output format specifiers? 

what if i want to have detailed output information, formatted as yaml or json?

File: pkg/console/console.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	if opts.Stderr != nil {
+		stderr = opts.Stderr
+	}
+}
+
+// Stdoutf formats according to a format specifier and writes the resulting string to stdout, swallowing any errors.
+// DO NOT USE Stdoutf for debug statements, warnings, notifications, or errors.
+// When an output format flag such as `-o json` is used, stdout should have a single parsable JSON object written to it, so that it can be piped and parsed by tools such as JQ.
+func Stdoutf(format string, a ...any) {
+	_, _ = fmt.Fprintf(stdout, format, a...)
```
Comment (Greg-At-Armory) [inline pkg/console/console.go:42]: is there somewhere any restriction what we pass into these functions ? if the requested "global" format specifier is json, would it work with the formatting code from api_error.go ?

Comment (fieldju) [reply pkg/console/console.go:42]: I answered this question in a Slack thread but I will put it here too for the curious.

We shouldn't be worrying about formatting errors into JSON or YAML.
In general with CLIs I have observed (kubectl, aws cli <- to name 2 big ones), errors are printed in plain text to stderr with non-0 exit codes
See this example from the aws cli (RED carrets == the previous command exited non-zero and green == previous command exited 0)

image

:point_up: following this pattern makes error handling much cleaner as you don't need to worry about what format a user requested and the stderr output logs to github logs and terminals even if you have the cli in pipes or `$()`

such as


```
#!/usr/bin/env bash

set -e
set -o pipefail

DEPLOYMENT_ID=$(armory deploy start -f scratch/deployment.yaml -o json | jq -r ".deploymentId")
```

File: internal/clierr/api_error.go (hunk 2 above)
Comment (fieldju) [reply internal/clierr/api_error.go:45]: I replied to this here: https://github.com/armory-io/armory-cli/pull/171#discussion_r1270795159

**Review by Greg-At-Armory — APPROVED** [review]

**Pull request merged**